src/validationinterface.cpp

```diff
@@ -4,49 +4,177 @@
+    boost::signals2::signal<void (const CTransactionRef &)> TransactionAddedToMempool;
+    boost::signals2::signal<void (const std::shared_ptr<const CBlock> &, const CBlockIndex *pindex, const std::vector<CTransactionRef>&)> BlockConnected;
+    boost::signals2::signal<void (const std::shared_ptr<const CBlock> &)> BlockDisconnected;
+    boost::signals2::signal<void (const CBlockLocator &)> SetBestChain;
+    boost::signals2::signal<void (const uint256 &)> Inventory;
+    boost::signals2::signal<void (int64_t nBestBlockTime, CConnman* connman)> Broadcast;
+    boost::signals2::signal<void (const CBlock&, const CValidationState&)> BlockChecked;
+    boost::signals2::signal<void (std::shared_ptr<CReserveScript>&)> ScriptForMining;
+    boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
+
+    CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    std::atomic<bool> fCallbacksRunning = ATOMIC_VAR_INIT(false);
+
+    void ProcessQueue() {
+        if (fCallbacksRunning.exchange(true)) {
+            return;
+        }
+
+        while (true) {
+            std::function<void (void)> callback;
+            {
+                LOCK(cs_callbacksPending);
+                if (callbacksPending.empty()) {
+                    // We must set fCallbacksRunning to false with
+                    // cs_callbacksRunning held - ensuring that there is no
+                    // race condition in which we might add something to the
+                    // queue, have the associated ProcessQueue call fail to get
+                    // the fCallbacksRunning lock, but have already given up on
+                    // doing more work in the previous ProcessQueue function.
+                    fCallbacksRunning = false;
+                    break;
+                }
+                callback = callbacksPending.front();
+                callbacksPending.pop_front();
+            }
+            try {
+                callback();
+            } catch (...) {
+                LogPrintf("Exception thrown in processing CValidationInterface callback\n");
+            }
+        }
+    }
+
+    void AddToProcessQueue(const std::function<void (void)>& func) {
+        if (!scheduler)
+            return;
+        {
+            LOCK(cs_callbacksPending);
+            callbacksPending.emplace_back(func);
+        }
+        scheduler->schedule(std::bind(&CMainSignalsInstance::ProcessQueue, this));
+    }
+};
 
 static CMainSignals g_signals;
 
+CMainSignals::CMainSignals() {
+    internals = new CMainSignalsInstance();
+}
+
+CMainSignals::~CMainSignals() {
+    delete internals;
```

Comment:
Sorry, but I'm going to insist that you maintain this commit intact (nevermind that it's likely the exact same patch you'd come up with if I hadn't shown you) forever in your PR history. Note: sarcasm.